### PR TITLE
[ENH] BEP001 - New entity: flip

### PR DESCRIPTION
**Pull request by agahkarakuzu**
*** 
### Headnote

A while ago we started #508, where you can find information on the overall purpose of `BEP001`. The PR #508 was closed as discussions called for some major revisions and to split `BEP001` pull requests into manageable parts.

*** 

Dear BIDS community, 

In #668, we proposed the common principle of `entity-linked file collections`, which requires addition of several new entities (`inv`, `flip`, `mt`) for the common use cases of quantitative MRI applications. In this **Part-3** of BEP001 pull requests, we would like to propose a new entity: 

|Entity | Corresponding metadata| Purpose |
| ---- | ---| --- |
| `flip`* | `FlipAngle` | To distinguish constituents of a `file-collection` if the files differ as a function of the `FlipAngle` acquisition parameter.|

\* We abstained from using its common abbreviation `fa`, as it may be confused with the `FA` (fractional anisotropy) term from `dwi` realm.

We set the requirement level of this entity to `MUST` for the applications falling within the definition of `entity-linked file collections`, **not for all the applications that may collect data at multiple `FlipAngle` values**. This is because splitting files may not be the most convenient/preferred approach for some applications, such as `ASL` #669. I took liberty to adjust the definition of the `echo` in that respect. 

I will be waiting for this PR to be discussed and resolved before sending PRs for the remaining entities `inv` and `mt`. 

On behalf of the BEP001 core team:

 Gilles de Hollander (@Gilles86), Alberto Lazari (@lazaral), Christophe Phillips (@ChristophePhillips), Kirstie Whitaker (@KirstieJane), Tibor Auer (@tiborauer). 

Best regards.

### CODE REVIEW

File: src/99-appendices/09-entities.md
Comment (tsalo) [inline src/99-appendices/09-entities.md:143]: This file is generated from a script. You need to add `flip` to the schema and use that to generate this file instead. [This](https://github.com/bids-standard/bids-specification/blob/580183fe1af767b4ea7ee3b0e9459713c0c6b658/src/schema/entities.yaml) is the file that needs to be added, and [this](https://github.com/bids-standard/bids-specification/blob/580183fe1af767b4ea7ee3b0e9459713c0c6b658/tools/bids_schema.py#L456-L457) is the script you need to run to regenerate the `09-entities.md` file.

**Review by tsalo — CHANGES_REQUESTED** [review]

File: src/99-appendices/09-entities.md
Comment (agahkarakuzu) [reply src/99-appendices/09-entities.md:143]: @tsalo so sorry that I missed that giant header comment. Fixing right away.

File: src/99-appendices/09-entities.md
Comment (agahkarakuzu) [inline src/99-appendices/09-entities.md:20]: @tsalo any idea why this is happening?

Comment (tsalo) [reply src/99-appendices/09-entities.md:20]: Argh, there's a bug in the generation code! The fix is buried in #324. I'm not sure if I should show you how to fix it here or open a separate PR with the fix...

Comment (tsalo) [reply src/99-appendices/09-entities.md:20]: For simplicity's sake, I'm going to ask you to make the change, if you're willing. At the following line, please replace `entity` with `entity_info['entity']`:

https://github.com/bids-standard/bids-specification/blob/580183fe1af767b4ea7ee3b0e9459713c0c6b658/tools/bids_schema.py#L227

Comment (agahkarakuzu) [reply src/99-appendices/09-entities.md]: @tsalo now it looks ok 👍

File: src/schema/entities.yaml
Comment (effigies) [inline src/schema/entities.yaml:167]:
```suggestion
  format: index
```

File: src/schema/entities.yaml
Comment (effigies) [inline src/schema/entities.yaml:114]:
```suggestion
  entity: flip
```

File: src/schema/entities.yaml
Comment (effigies) [inline src/schema/entities.yaml:122]:
```suggestion
  format: index
```

**Review by effigies — APPROVED** [review]
LGTM. Cleaning yaml lint...

**Comment by agahkarakuzu** [comment]
Thank you @effigies! Looks like all the CI requirements are satisfied now.

File: src/99-appendices/09-entities.md
Comment (tsalo) [inline src/99-appendices/09-entities.md:141]: I think this revised definition is worded a little confusingly. Rather than "constituents of an entity-linked file collection", I would say "files that are part of an entity-linked file collection". Instead of "differ as a function of `EchoTime`", I think you could say "have different `EchoTime`s.

Comment (tsalo) [reply src/99-appendices/09-entities.md:141]: Also, I think the definition should go before the usage and warning. Something like "The `_echo-<index>` key/value pair represents the `EchoTime` metadata field." Then you can go into when and how it should be used.

File: src/schema/entities.yaml
Comment (tsalo) [inline src/schema/entities.yaml:107]: I think this revised definition is worded a little confusingly. Rather than "constituents of an entity-linked file collection", I would say "files that are part of an entity-linked file collection". Instead of "differ as a function of `EchoTime`", I think you could say "have different `EchoTime`s.

Comment (tsalo) [reply src/schema/entities.yaml:107]: Also, I think the definition should go before the usage and warning. Something like "The `_echo-<index>` key/value pair represents the `EchoTime` metadata field." Then you can go into when and how it should be used.

File: src/99-appendices/09-entities.md
Comment (tsalo) [reply src/99-appendices/09-entities.md:141]: Okay I just realized I commented on the rendered file instead of the source one 😞. Resolving this and commenting in the right place.

File: src/schema/entities.yaml
Comment (agahkarakuzu) [reply src/schema/entities.yaml:107]: I agree, re-wording this now.

**Comment by effigies** [comment]
I think rebasing on master should resolve the CI failures.

**Comment by sappelhoff** [comment]
@agahkarakuzu do you have the "allow edits from maintainers" box in the GitHub user interface ticked? I cannot seem to push to your fork's branch. (wanted to rebase on master)

**Comment by agahkarakuzu** [comment]
@sappelhoff I don't see that box on the PR page, probably because I am not an owner but a member. I'll try to solve that.

**Comment by effigies** [comment]
@agahkarakuzu It's a checkbox when you create a new PR:

Screen Shot 2020-11-19 at 09 56 41

I don't know that it can be changed after the fact.

**Comment by agahkarakuzu** [comment]
@effigies it was not available to me while I was creating the PR neither. I can rebase to a master clone branch on our fork, would that work?

**Comment by agahkarakuzu** [comment]
@effigies I gave it a try, hope that it won't mess it up.

**Review by effigies — APPROVED** [review]

**Comment by effigies** [comment]
Last semantic change was Monday, so this can be merged this coming Monday if we get another review.

Recommend squashing to keep things clean.

**Review by tsalo — APPROVED** [review]

**Review by sappelhoff — APPROVED** [review]

**Comment by sappelhoff** [comment]
Aaand it's in! Thanks a lot everyone and especially @agahkarakuzu  for preparing the PR.

**Comment by agahkarakuzu** [comment]
Thank you @sappelhoff I will be sending PRs for the remaining entities this week, I think this was a good starting point towards merging bep001!

**Pull request merged**
